dashboard/modules/state/state_head.py

```diff
@@ -87,55 +95,125 @@ async def _update_agent_stubs(self, change: Change):
+            )
+            return self._reply(
+                success=True,
+                error_message="",
+                result=result.result,
+                partial_failure_warning=result.partial_failure_warning,
+            )
+        except DataSourceUnavailable as e:
+            return self._reply(success=False, error_message=str(e), result=None)
 
     @routes.get("/api/v0/workers")
     async def list_workers(self, req) -> aiohttp.web.Response:
-        data = await self._state_api.list_workers(option=self._options_from_req(req))
-        return self._reply(success=True, message="", result=data)
+        try:
+            result = await self._state_api.list_workers(
+                option=self._options_from_req(req)
+            )
+            return self._reply(
+                success=True,
+                error_message="",
+                result=result.result,
+                partial_failure_warning=result.partial_failure_warning,
+            )
+        except DataSourceUnavailable as e:
+            return self._reply(success=False, error_message=str(e), result=None)
 
     @routes.get("/api/v0/tasks")
     async def list_tasks(self, req) -> aiohttp.web.Response:
-        data = await self._state_api.list_tasks(option=self._options_from_req(req))
-        return self._reply(success=True, message="", result=data)
+        try:
+            result = await self._state_api.list_tasks(
+                option=self._options_from_req(req)
+            )
+            return self._reply(
+                success=True,
+                error_message="",
+                result=result.result,
+                partial_failure_warning=result.partial_failure_warning,
+            )
+        except DataSourceUnavailable as e:
+            return self._reply(success=False, error_message=str(e), result=None)
 
     @routes.get("/api/v0/objects")
     async def list_objects(self, req) -> aiohttp.web.Response:
-        data = await self._state_api.list_objects(option=self._options_from_req(req))
-        return self._reply(success=True, message="", result=data)
+        try:
+            result = await self._state_api.list_objects(
+                option=self._options_from_req(req)
+            )
+            return self._reply(
+                success=True,
+                error_message="",
+                result=result.result,
+                partial_failure_warning=result.partial_failure_warning,
+            )
+        except DataSourceUnavailable as e:
+            return self._reply(success=False, error_message=str(e), result=None)
 
     @routes.get("/api/v0/runtime_envs")
     @dashboard_optional_utils.aiohttp_cache
     async def list_runtime_envs(self, req) -> aiohttp.web.Response:
-        data = await self._state_api.list_runtime_envs(
-            option=self._options_from_req(req)
-        )
-        return self._reply(success=True, message="", result=data)
+        try:
```

Comment:
Seems this code template is repeated multiple times. Could we create a helper function?